menpofit/dpm/base.py

```diff
@@ -0,0 +1,1062 @@
+                s['box_y2'] = [p['box_y2'][i]]
+                s['box_x2'] = [p['box_x2'][i]]
+                split_pos[k].append(dict(s))
+        return split_pos
+
+    def _ln2box(self, pos):
+        # Converts the points into bounding boxes.
+        for i, p in enumerate(pos):
+            parents_lns = self._get_parents_lns(p['gmixid'])
+            lengths = np.linalg.norm(abs(p['pts'][1:] - p['pts'][parents_lns[1:]]), axis=1)
+            boxlen = np.percentile(lengths, 80, interpolation='midpoint')
+            assert(boxlen > 3)  # ensure that boxes are 'big' enough.
+            _t = np.clip(p['pts'] - 1 - boxlen/2, 0, np.inf) # -1 for matlab indexes
+            p['box_x1'] = np.copy(_t[:, 0])
+            p['box_y1'] = np.copy(_t[:, 1])
+            _t = p['pts'] - 1 + boxlen/2  # no check for boundary, -1 for matlab indexes
+            p['box_x2'] = np.copy(_t[:, 0])
+            p['box_y2'] = np.copy(_t[:, 1])
+        return pos
+
+    def _init_model(self, pos_, sbin):
+        areas = np.empty((len(pos_),))
+        for i, el in enumerate(pos_):
+            areas[i] = (el['box_x2'][0] - el['box_x1'][0] + 1) * (el['box_y2'][0] - el['box_y1'][0] + 1)
+        areas = np.sort(areas)
+        area = areas[np.floor(areas.size * 0.2)]  # Pick the 20th percentile area
+        nw = np.sqrt(area)
+
+        im = hog(Image(np.zeros((1, 30, 30))), mode='sparse', algorithm='zhuramanan')  # Calculating HOG features size
+        siz = [im.pixels.shape[0], round(nw/sbin), round(nw/sbin)]
+
+        d = dict()  # deformation
+        d['w'] = 0  # bias
+        d['i'] = 0
+        d['anchor'] = np.zeros((3,))
+
+        f = dict()  # filter
+        f['w'] = np.empty(siz)
+        f['i'] = 1
+
+        c = dict()
+        c['filter_ids'] = [0]
+        c['def_ids'] = [0]
+        # c['parent'] = -1
+        # c['anchors'] = [np.zeros((3,))]
+        c['tree'] = Tree(np.array([0]), root_vertex=0, skip_checks=True)  # Independent part is a single node tree.
+
+        model = dict()
+        model['defs'] = [d]
+        model['filters'] = [f]
+        model['components'] = [c]
+
+        model['maxsize'] = siz[1:]
+        model['len'] = 1 + np.prod(siz)
+        model['interval'] = 10
+        model['sbin'] = sbin
+
+        model = self._poswarp(model, pos_)
+        return Model.model_from_dict(model)
+        # c = model['components'][0]
+        # c['bias'] = d['w']  # todo: dont think this is needed
+
+    def _poswarp(self, model, pos):
+        # Update independent part model's filter by averaging its hog feature.
+        warped = self._warppos(model, pos)
+        s = model['filters'][0]['w'].shape  # filter size
+        num = len(warped)
+        feats = np.empty((np.prod(s), num))
+
+        for c, im in enumerate(warped):
+            feat = hog(im, mode='sparse', algorithm='zhuramanan', cell_size=model['sbin'])
+            feats[:, c] = feat.pixels.flatten()
+
+        w = np.mean(feats, 1)
```

Comment:
This needs ``axis=1`` because it won't work on Python 3 where kwargs are required to be explicit (rightly so).